File: src/main/scala-2.11/com/lei/leetcode/P606/Solution.java
package com.lei.leetcode.P606;

import com.lei.leetcode.TreeNode;

public class Solution {
    public String tree2str(TreeNode t) {
        if (t == null) return "";

        String result = t.val + "";

        String left = tree2str(t.left);
        String right = tree2str(t.right);

        if (left.equals("") && right.equals("")) return result;
        if (left.equals("")) return result + "()" + "(" + right + ")";
        if (right.equals("")) return result + "(" + left + ")";
        return result + "(" + left + ")" + "(" + right + ")";
    }
}

File: src/main/scala-2.11/com/lei/leetcode/P606/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def tree2str(self, t):
        if t is None:
            return ''
        val = str(t.val)
        if t.left is None and t.right is None:
            return val
        elif t.left is not None and t.right is None:
            return val + '(' + self.tree2str(t.left) + ')'
        else:
            return val + '(' + self.tree2str(t.left) + ')(' + self.tree2str(t.right) + ')'
